Add Keyboard Lock article (#3557)
GoogleChrome/web.dev: src/site/content/en/blog/keyboard-lock/index.md

---
layout: post
title: Capture Keys with the Keyboard Lock API
subhead: Provide an immersive, full screen experience for a variety of use cases including interactive websites, games, and remote desktop or application streaming.
tags:

>   - blog

Author: ```suggestion
  - blog
  - capabilities
```

authors:
  - thomassteiner
description: >
  The Keyboard Lock API allows websites to capture keys that are normally reserved by the underlying operating system. It is intended for web applications that provide a full screen immersive experience (like games or remote access apps).

> date: 2020-07-23

Author: ```suggestion
date: 2020-07-28
```

# updated:

> hero:

Author: ```suggestion
hero: hero.jpg
```

---
layout: post
title: Capture Keys with the Keyboard Lock API
subhead: Provide an immersive, full screen experience for a variety of use cases including interactive websites, games, and remote desktop or application streaming.
tags:
  - blog
  - capabilities
authors:
  - thomassteiner
description: >
  The Keyboard Lock API allows websites to capture keys that are normally reserved by the underlying operating system. It is intended for web applications that provide a full screen immersive experience (like games or remote access apps).
date: 2020-07-28
# updated:
hero: hero.jpg
---

> With more and more users spending most of their time in the browser, richly interactive websites, games, and remote desktop or application streaming experiences strive to provide an immersive, full screen experience. To accomplish this, sites need access to special keys and keyboard shortcuts while they are in full screen mode, so that they can be used for navigation, menus, or gaming functionality. Some examples of the keys that may be required are <kbd>Esc</kbd>, <kbd>Alt</kbd>&nbsp;+&nbsp;<kbd>Tab</kbd>, <kbd>Cmd</kbd>&nbsp;+&nbsp;<kbd>`</kbd>, and <kbd>Ctrl</kbd>&nbsp;+&nbsp;<kbd>N</kbd>.

Author: ```suggestion
With more and more users spending most of their time in the browser, richly interactive websites, games, remote desktop streaming, and application streaming strive to provide an immersive, full screen experience. To accomplish this, sites need access to special keys and keyboard shortcuts while they are in full screen mode, so that they can be used for navigation, menus, or gaming functionality. Some examples of the keys that may be required are <kbd>Esc</kbd>, <kbd>Alt</kbd>&nbsp;+&nbsp;<kbd>Tab</kbd>, <kbd>Cmd</kbd>&nbsp;+&nbsp;<kbd>`</kbd>, and <kbd>Ctrl</kbd>&nbsp;+&nbsp;<kbd>N</kbd>.
```

#### Capturing all keys

The following example captures all key presses.

```js
navigator.keyboard.lock();
```

#### Capturing specific keys

The following example captures the <kbd>W</kbd>, <kbd>A</kbd>, <kbd>S</kbd>, and <kbd>D</kbd> keys. It captures these keys regardless of which modifiers are used with the key press. Assuming a US QWERTY layout, registering `"KeyW"` ensures that <kbd>W</kbd>, <kbd>Shift</kbd>&nbsp;+&nbsp;<kbd>W</kbd>, <kbd>Control</kbd>&nbsp;+&nbsp;<kbd>W</kbd>, <kbd>Control</kbd>&nbsp;+&nbsp;<kbd>Shift</kbd>&nbsp;+&nbsp;<kbd>W</kbd>, and all other key modifier combinations with <kbd>W</kbd> are sent to the app. The same applies to `"KeyA"`, `"KeyS"`, and `"KeyD"`.

```js
await navigator.keyboard.lock([
  "KeyW",
  "KeyA",
  "KeyS",
  "KeyD",
]);
```

> If `lock()` is called multiple times without an intervening call to `unlock()` (see [below](#unlocking-the-keyboard)), then only the key codes specified in the last call will be in effect. If a second call to `lock()` is made before the first one has finished, then the first one will be rejected with an `AbortError`.

Author: ```suggestion
You can react on captured key presses for example in the `onkeydown` event:

```js
document.addEventListener('keydown', (e) => {
  if ((e.code === 'KeyA') && !(event.ctrlKey || event.metaKey)) {
    // Do something when the 'A' key was pressed, but only
    // when not in combination with the command or control key.
  }
});
```

If `lock()` is called multiple times without an intervening call to `unlock()` (see [below](#unlocking-the-keyboard)), then only the key codes specified in the last call will be in effect. If a second call to `lock()` is made before the first one has finished, then the first one will be rejected with an `AbortError`.

#### Capturing all keys

The following example captures all key presses.

```js
navigator.keyboard.lock();
```

#### Capturing specific keys

The following example captures the <kbd>W</kbd>, <kbd>A</kbd>, <kbd>S</kbd>, and <kbd>D</kbd> keys. It captures these keys regardless of which modifiers are used with the key press. Assuming a US QWERTY layout, registering `"KeyW"` ensures that <kbd>W</kbd>, <kbd>Shift</kbd>&nbsp;+&nbsp;<kbd>W</kbd>, <kbd>Control</kbd>&nbsp;+&nbsp;<kbd>W</kbd>, <kbd>Control</kbd>&nbsp;+&nbsp;<kbd>Shift</kbd>&nbsp;+&nbsp;<kbd>W</kbd>, and all other key modifier combinations with <kbd>W</kbd> are sent to the app. The same applies to `"KeyA"`, `"KeyS"`, and `"KeyD"`.

```js
await navigator.keyboard.lock([
  "KeyW",
  "KeyA",
  "KeyS",
  "KeyD",
]);
```

> You can react on captured key presses for example in the `onkeydown` event:

Reviewer: ```suggestion
You can respond to captured key presses using keyboard events. 
For example this code uses the `onkeydown` event:
```
